src/protocol/client.c

```diff
@@ -143,191 +140,190 @@ protocolClientNew(const String *name, const String *service, IoRead *read, IoWri
 void
-protocolClientWriteCommand(ProtocolClient *this, const ProtocolCommand *command)
+protocolClientDataEndGet(ProtocolClient *const this)
 {
     FUNCTION_LOG_BEGIN(logLevelTrace);
         FUNCTION_LOG_PARAM(PROTOCOL_CLIENT, this);
-        FUNCTION_LOG_PARAM(PROTOCOL_COMMAND, command);
     FUNCTION_LOG_END();
 
-    ASSERT(this != NULL);
-    ASSERT(command != NULL);
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        PackRead *response = pckReadNew(this->pub.read);
+        ProtocolMessageType type = (ProtocolMessageType)pckReadU32P(response);
 
-    // Write out the command
-    ioWriteStrLine(protocolClientIoWrite(this), protocolCommandJson(command));
-    ioWriteFlush(protocolClientIoWrite(this));
+        protocolClientError(this, type, response);
 
-    // Reset the keep alive time
-    this->keepAliveTime = timeMSec();
+        pckReadEndP(response);
+
+        CHECK(type == protocolMessageTypeDataEnd);
+    }
+    MEM_CONTEXT_TEMP_END();
 
     FUNCTION_LOG_RETURN_VOID();
 }
 
 /**********************************************************************************************************************************/
-const Variant *
-protocolClientExecute(ProtocolClient *this, const ProtocolCommand *command, bool outputRequired)
+void
+protocolClientCommandPut(ProtocolClient *const this, ProtocolCommand *const command)
 {
     FUNCTION_LOG_BEGIN(logLevelTrace);
         FUNCTION_LOG_PARAM(PROTOCOL_CLIENT, this);
         FUNCTION_LOG_PARAM(PROTOCOL_COMMAND, command);
-        FUNCTION_LOG_PARAM(BOOL, outputRequired);
     FUNCTION_LOG_END();
 
     ASSERT(this != NULL);
     ASSERT(command != NULL);
 
-    protocolClientWriteCommand(this, command);
+    // Put command
+    protocolCommandPut(command, this->write);
+
+    // Reset the keep alive time
+    this->keepAliveTime = timeMSec();
 
-    FUNCTION_LOG_RETURN_CONST(VARIANT, protocolClientReadOutput(this, outputRequired));
+    FUNCTION_LOG_RETURN_VOID();
 }
 
 /**********************************************************************************************************************************/
-void
-protocolClientNoOp(ProtocolClient *this)
+PackRead *
+protocolClientExecute(ProtocolClient *const this, ProtocolCommand *const command, const bool resultRequired)
 {
     FUNCTION_LOG_BEGIN(logLevelTrace);
         FUNCTION_LOG_PARAM(PROTOCOL_CLIENT, this);
+        FUNCTION_LOG_PARAM(PROTOCOL_COMMAND, command);
+        FUNCTION_LOG_PARAM(BOOL, resultRequired);
     FUNCTION_LOG_END();
 
     ASSERT(this != NULL);
+    ASSERT(command != NULL);
 
-    MEM_CONTEXT_TEMP_BEGIN()
-    {
-        protocolClientExecute(this, protocolCommandNew(PROTOCOL_COMMAND_NOOP), false);
-    }
-    MEM_CONTEXT_TEMP_END();
+    // Put command
+    protocolClientCommandPut(this, command);
 
-    FUNCTION_LOG_RETURN_VOID();
+    // Read result if required
+    PackRead *result = NULL;
+
+    if (resultRequired)
+        result = protocolClientDataGet(this);
+
+    // Read response
+    protocolClientDataEndGet(this);
```

Comment:
> we perform a protocolClientDataGet(this->client) at line 85 and if there was not a result, then we do a DataEndGet() - so that seems to imply we would need to have an else in this code (see below). But maybe src/storage/remote/read.c is wrong?

We do DataEndGet() here if there was no file to read. If there *is* a file to read we do DataEndGet() at line 155.

> I can't seem to find the rules on how the client protocol Get/Put are supposed to work - like why is there a protocolClientDataPut() but no protocolClientDataEndPut()?

Basically because it is not needed in most cases. If we add this we'd have to check it on the server side everywhere, which doesn't seem worth it. Eventually each command will have metadata so we know whether it sends or receives data, but for now I wanted to keep it similar to how it was in the prior protocol.